core/ir/opnd.h

```diff
@@ -1777,6 +1777,34 @@ opnd_t
 opnd_create_base_disp_aarch64(reg_id_t base_reg, reg_id_t index_reg,
                               dr_extend_type_t extend_type, bool scaled, int disp,
                               dr_opnd_flags_t flags, opnd_size_t size);
+DR_API
+/**
+ * Creates an unsigned immediate integer operand for condition defined by enum
+ * constant \p cond of type #dr_pred_type_t.
+ * \note AArch64-only.
+ */
+opnd_t
+opnd_create_cond(int cond);
+
+DR_API
+/**
+ * Assumes the operand is an unsigned immediate 4-bit integer and returns a
+ * #dr_pred_type_t constant corresponding to the value of the operand.
+ * \note AArch64-only.
+ */
+int
+opnd_get_cond(opnd_t opnd);
+
+DR_API
+/**
+ * Assumes the operand is an unsigned immediate 4-bit integer and returns an
+ * operand which contains inverted condition (see #dr_pred_type_t).
+ * \note Does not support AL and NV conditions.
+ * \note AArch64-only.
+ */
+opnd_t
+opnd_invert_cond(opnd_t opnd);
```

Comment:
I think this needs some thought as to how to integrate with existing interfaces and code, especially with ARM code.

The API already has `instr_invert_predicate`. I would agree it seems less of an instr-level operation, but it would be confusing to have two different sets of interfaces.

The API also already has predicates passed as dr_pred_type_t to things like `XINST_CREATE_jump_cond` in arm/instr_create.h.  If a new type of operand, even if it's an immediate underneath is created, it would be best to use it everywhere.  Whatever is decided, it seems best to have the same approach for all platforms, rather than special AArch64-only functions that don't match ARM functions yet do the same thing.  That makes for confusing interfaces and complex cross-platform clients.

Reply:
It would be a very big change to move all the condition-related interfaces that currently exist on the level of instructions to the level of operands. I don't think I know enough about other platforms to come up with a uniform solution, and I don't know if such change actually makes sense for x86 (after all `instr_invert_predicate` was create for some reason).

Anyway, I'd like to do it in steps rather than in a single big pull request which is very difficult to merge. At least some guidance, suggestions or ideas would be helpful.

I would start with #4502, but again as a separate patch.

Reply:
Right now I could remove `opnd_invert_cond` and use `instr_invert_predicate` around `instr_create_...` to flip the operand, however I'm not sure if `instr_invert_predicate` is actually implemented for AArch64.

Reply:
See https://dynamorio.org/dynamorio_docs/API_BT.html#sec_predication which lists one key current use of predicates/conditions:
+ An instruction-level modifier that is not an operand: `instr_{get,set}_predicate`; `INSTR_PRED` short form.  While perhaps this could have been an operand (along with other things like condition code effects: tradeoffs with matching ISA conventions), it is not.  Yet we want to be able to invert these predicates too.  The current `instr_invert_predicate` does not actually operate on `instr_t`: it operates on `dr_pred_type_t` and thus operates on the instr-level predicates as well as operand-level.

Another current use, similar to the macros added here:
+ `XINST_CREATE_jump_cond()`.  This, on x86, arm, and aarch64, takes in a `dr_pred_type_t`-typed value.  I would think that the signature/operation of this instruction should match things like the `INSTR_CREATE_ccmn` added here.  One shouldn't take a `dr_pred_type_t` while the other takes an `opnd_t`.  Maybe the argument could be made that we should change the `XINST_CREATE_jump_cond` to take an `opnd_t` of this new type: though note that the condition turns into different opcodes on x86, the instruction-level predicate field on arm, and an operand on aarch64, so they are not all treating it as an operand.  See also other high-level discussions on what should constitute an operand versus an opcode or sub-opcode: https://github.com/DynamoRIO/dynamorio/pull/4386#discussion_r462356954.  Perhaps these aarch64 comparison should be using separate opcodes, rather than operands that change behavior which goes against what operands are supposed to do.  Having a single opcode doing all these different comparisons is sort of like having one opcode "arithmetic" with an operand that says whether it's an add or a subtract or whatnot, instead of separate add and subtract opcodes.

Reply:
Continuing in the thread of the last few sentences: xref other issues where we want to split up the current aarch64 opcodes which have behavior-controlling integer operands which are perhaps better as separate opcodes rather than operands: #4388, #4393.

Reply:
> One shouldn't take a dr_pred_type_t while the other takes an opnd_t.

I think it would be much better if all `INSTR_CREATE_` macros would take values instead of operands (e.g. `DR_REG_` constants instead of `opnd_create_reg(...)`), but most of such macros take `opnd_t`, that is why I implemented new macros in such a way -- to follow the common approach. I believe such choice was made for better portability of macros across different architectures. So, I agree that `XINST_CREATE_jump_cond` is an exception, from this point of view.

Conditional operands in AArch64 work differently from other platforms, that is why I thought having an operand-level function for AArch64 was a better idea than extending `instr_invert_predicate` (I suggest to discuss it in #4504).

Maybe we shouldn't make inherently different architectures conform to some common notion of how instructions "should" work?

Reply:
> Maybe we shouldn't make inherently different architectures conform to some common notion of how instructions "should" work?

The whole point of an IR is to abstract away underlying machine encoding quirks.  It is a major burden on client writers if you impose on them different IR conventions for different architectures.  Plus, this predicate here is not anything inherently different from other architectures: what's different is that the ISA *manual* presents it as an operand rather than part of the opcode, while other architectures would present it as part of the opcode with zero changes in how it works.  Yes, some things are fundamentally different, like universal predication on ARM, but we should make all the other things as similar as we can in the IR.

I would vote for not having any "magic behavior modifiers" as integer operands anywhere in DR's IR: they should be part of the opcode.  The only integer operand should be an actual integer value, not an encoding of how to change the behavior.  The behavior comes from the opcode, not the operands.  So this CCMN and CCMP would have their opcodes split up.  Just like the SYS opcode should be split up as discussed elsewhere.  Just like on ARM we split up the SIMD opcodes and have `OP_vld2_lane_8` instead of just one `OP_vld2` with magic operands encoding behavior and then special operand types for each of those integer-encoded operands.

@AssadHashmi WDYT about splitting up CCMN and CCMP?

Reply:
> I think it would be much better if all INSTR_CREATE_ macros would take values instead of operands (e.g. DR_REG_ constants instead of opnd_create_reg(...)),

This breaks down for complex operands like memory references.  Some instructions can take either a register or a memory reference for the same operand slot.

Reply:
> WDYT about splitting up CCMN and CCMP?

Why only these instructions? All conditional instructions in AArch64 (all mentioned in this PR) have `cond` operand.

Also, in my opinion, inventing non-documented opcodes may create extra confusion. If some notion is not common to the rest of the architectures it's not the reason to disallow low-level access _as documented_. I think in DynamoRIO, when we work at the level of instructions and operands, we don't really have the benefit of "write once run everywhere". In my understanding, the point of the tool is to allow working on the lowest possible level.

Reply:
>  The behavior comes from the opcode, not the operands.

How about compare and swap instructions (e.g. CAS)?  The behaviour depends on the value of the register being equal to the value loaded from memory. It is similar to the conditional instructions, although in the latter case the comparison has happened _before_.

Reply:
It might be useful to think of the pseudocode that specifies instruction behaviors. CCMP is specified by one pseudocode function, and the operands are inputs to that function. That single behavior function acts uniformly on all values of the operands. So it makes sense to treat it as one instruction, rather than single out one operand and split it into one instruction for each value of that operand.

The situation with SYS instructions is very different (similar to CDP instructions in 32-bit Arm) - they really are separate functions, selected by a function code, and it does make much more sense to split them out, at least when their specific functions are being differently handled by DynamoRIO.

Reply:
> @AssadHashmi WDYT about splitting up CCMN and CCMP?

If I understand correctly we have 3 choices:

1 - Create a new operand for `cond`.
2 - Split out the opcodes into specific `cond` types, e.g. `ccmp_eq`, `ccmp_ne` etc.
3 - Use existing `dr_pred_type_t` as in `XINST_CREATE_jump_cond` (alias for `INSTR_CREATE_bcond`).

1 requires a non-trivial amount of work for a type which has essentially the same intent as `dr_pred_type_t`.

2 is not consistent with AArch64's existing codec. We may run into problems as I did in https://github.com/DynamoRIO/dynamorio/pull/4386

3 would be my choice. It makes use of an existing type with the same intent which has already been implemented and does not involve a lot of codec reworking.
e.g.
```
#define INSTR_CREATE_ccmp(dc, Rn, Op, nzcv, cond) \
    (INSTR_PRED(instr_create_0dst_3src(dc, OP_ccmp, Rn, Op, nzcv), (cond)))
```
Which users can call with:
```
INSTR_CREATE_ccmp(dc, opnd_create_reg(DR_REG_X1),
        opnd_create_immed_int(0b01010, OPSZ_5b),
        opnd_create_immed_int(0b1111, OPSZ_4b), DR_PRED_EQ);
```

Reply:
@AssadHashmi highlights what is perhaps the main issue here: using a simple and consistent set of abstractions and types in the IR.  If we take each encoded behavior immediate integer that AArch64 presents as an operand (but I would call a sub-opcode and another ISA might well make part of the opcode, such as x86 conditional moves) and place it as a separate operand in DR's IR, we then need to make a new operand type with a new set of rules and enums and handling code and introduce a new concept to tool writers for what could logically fit into existing concepts.  We could end up with a dozen new operand types without much benefit but incurring non-trivial costs.  Using an existing IR concept keeps the IR simpler and cleaner and more consistent.

> we don't really have the benefit of "write once run everywhere".

We can make it as easy as possible, though.  It does make a difference: each IR choice that lets a tool or analysis library be more platform-agnostic makes tool writing easier.

> In my understanding, the point of the tool is to allow working on the lowest possible level.

None of any of the discussion here has any bearing on whether a tool writer can generate any precise machine code desired: we're only talking about how to represent ISA concepts in the DR IR.  None of the choices affect *whether* the ISA concept can be represented or whether the tool writer can generate it.  And while the IR should be more abstracted and cross-platform, we can and do have layers that are closer to the assembly conventions of that ISA: the INSTR_CREATE_ macros try to look more like assembly, disassembly has both an abstracted DR format and an ISA format, and we'd love to have an even closer layer where assembler strings are converted into IR (can't seem to find it in the tracker...#4510) when writing code for a specific machine.  But code analyzing and manipulating at the IR level should have as few types and special operand concepts as possible.